### PR TITLE
Fiber aware shared mutex

### PR DESCRIPTION
**Pull request by rgarc**
This PR copies the `std::shared_mutex` from the [standard library](https://github.com/gcc-mirror/gcc/blob/36eec7995b4d53083c3ee7824bd765b5eba8b1a1/libstdc%2B%2B-v3/include/std/shared_mutex#L265-L408), but uses synchronization primitives that are fiber aware (`boost::fibers::mutex` and `boost::fibers::condition_variable` instead of the `std` variants).

Problem:
- We would like to use a reader-writer lock in the context of a boost
  fiber.
- Boost.Fiber does not provide one.

Solution:
- Copy the `shared_mutex` implementation from the standard library but
  replace the blocking synchronization primitives with fiber-aware ones
  that yield the current fiber instead of blocking.

### CODE REVIEW

File: include/monad/fiber/shared_mutex.hpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            ++state_;
+            return true;
+        }
+        return false;
+    }
+
+    void unlock_shared()
+    {
+        std::lock_guard<boost::fibers::mutex> guard(mutex_);
+        MONAD_ASSERT(readers() > 0);
```
Comment (sheapwilliams) [inline include/monad/fiber/shared_mutex.hpp:97]: This call should only be called paired with `lock_shared()`, so I understand that this assert really protects against being called when paired with `lock()`, yea?

If so, then is this just protecting against programming errors?  Is `MONAD_ASSERT` the right thing for detecting programming errors?

File: src/monad/fiber/test_shared_mutex.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+struct SharedMutex : public testing::Test
+{
+};
+
+using MutexTypes =
+    ::testing::Types<DumbMutex, monad::shared_mutex, std::shared_mutex>;
+
+TYPED_TEST_SUITE(SharedMutex, MutexTypes);
+
+TYPED_TEST(SharedMutex, DISABLED_simple_bench)
```
Comment (sheapwilliams) [inline src/monad/fiber/test_shared_mutex.cpp]: Why disabled?  What is the follow-on plan?

File: src/monad/fiber/test_shared_mutex.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                    }
+                }
+                for (auto &fiber : fibers) {
+                    fiber.join();
+                }
+            });
+        }
+    }
+
+    EXPECT_EQ(counter.get(), 34);
```
Comment (sheapwilliams) [inline src/monad/fiber/test_shared_mutex.cpp:146]: Curious about getting better coverage on this...but this is probably fine for now.

File: include/monad/fiber/shared_mutex.hpp (hunk 1 above)
Comment (rgarc) [reply include/monad/fiber/shared_mutex.hpp:97]: I see. After talking about asserts yesterday in the run loop meeting, it's clear this should be a debug assert.

File: src/monad/fiber/test_shared_mutex.cpp (hunk 2 above)
Comment (rgarc) [reply src/monad/fiber/test_shared_mutex.cpp]: I'm actually not sure what we want to do for benchmark tests. Motivation here was to not hog up CI, but I was also not sure if this warranted being a separate binary.

File: src/monad/fiber/test_shared_mutex.cpp (hunk 3 above)
Comment (rgarc) [reply src/monad/fiber/test_shared_mutex.cpp:146]: I was also puzzled at coming up with a reasonable testing strategy for this since all the interesting edge cases are more likely to arise from using schedulers other than the default round robin.
